design/deployment-strategies.md

```diff
@@ -0,0 +1,212 @@
+      postRouteTafficCheck:
+        timeout: 60m
+        samplingInterval: 5m
+        steps:          
+        - script: echo monitor app health        
+      on:
+        failure:
+          steps:
+          - script: echo rollback, cleanup..     
+        success:
+          steps:
+          - script: echo checks passed...
+```
+
+For example, deploy a web app using blue-green strategy.
+
+```yaml
+jobs:
+- deployment:
+  environment: musicCarnivalProd.musicCarnivalWeb
+  pool:
+    name: musicCarnivalProdPool
+  strategy:                   
+    blueGreen:    
+      deploy:              
+        steps:                                   
+        - task: AzureWebApp@1
+          inputs:
+            appName: 'musicCarnivalWeb'
+            package: '$(System.DefaultWorkingDirectory)/**/*.zip' 
+            slotName: $(environment.staging)                        # deploy to green
+      routeTraffic:
+        delay: 60m
+        steps:
+        - task: AzureAppServiceManage@0
+          inputs:
+            Action: 'Swap Slots'                                    # swap green with blue
+            WebAppName: 'musicCarnivalWeb'
+            sourceSlot: $(environment.staging)
+            destinationSlot: $(environment.prod)
+      postRouteTafficCheck:
+        timeout: 60m
+        samplingInterval: 5m
+        steps:          
+        - script: echo monitor app health        
+      on:
+        failure:
+          steps:
+          - task: AzureAppServiceManage@0
+            inputs:
+              Action: 'Swap Slots'
+              WebAppName: 'musicCarnivalWeb'
+              sourceSlot: $(environment.prod)
+              destinationSlot: $(environment.staging)    
+        success:
+          steps:
+          - script: echo checks passed...
+```
+
+The example task usage above uses the explicit references to **staging** and **prod** slots from an environment variable. Typically built-in tasks such as AppService publishes the target type i.e, **Prod** or **staging** in other words **blue** or **green** information as a metadata on the resource in the environment context. With this, the target type whether blue or green information is persisted, updated and/or retrieved from the enviroment context and the workflow doesn't have to be edited for deployment. We also want to provide simplified commands/APIs to persist/update the slot information applied to the example above. 
+
+This example, can be scaled for any resource for example, **VMs** or **VMSS** or **K8S**. 
+
+**Canary**: reduce the risk by slowly rolling out the change to a small subset of users. 
+As you gain more confidence in the new version, you can start releasing it to more servers in your infrastructure
+and routing more users to it. 
+
+Consider the following Azure Pipelines pipeline defined in YAML
+
+```yaml
+jobs:
+- deployment:
+  environment: musicCarnivalProd
+  pool:
+    name: musicCarnivalProdPool 
+  strategy:                 
+    canary:     
```

Comment:
Can you please specify how each of the increments input will map to a new iteration and the list of variables that will be available for the tasks in the iteration.

Reply:
Each iteration will include the following variables $(increments), $(strategy), and $(action) in case of canary.  These variables can be used/mapped to the step inputs in case of k8s-deploy